File: backend/models/review.py
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field

Provider = Literal["kling", "comfy_video", "comfy_image"]
ReviewStatus = Literal[
    "pending_review", "approved", "dispatched", "completed", "failed", "discarded"
]


class ReviewItemCreate(BaseModel):
    source_image_path: str
    prompt: str
    provider: Provider
    workflow_name: Optional[str] = None
    settings: Dict = Field(default_factory=dict)


class ReviewCreateRequest(BaseModel):
    items: List[ReviewItemCreate] = Field(min_length=1)
    batch_id: Optional[str] = None


class ReviewCreateResponse(BaseModel):
    batch_id: str
    request_ids: List[str]


class ReviewRequestItem(BaseModel):
    id: str
    batch_id: str
    source_image_path: str
    original_prompt: str
    prompt: str
    provider: str
    workflow_name: Optional[str] = None
    settings: Dict = Field(default_factory=dict)
    status: str
    execution_id: Optional[str] = None
    result_path: Optional[str] = None
    error: Optional[str] = None
    project_id: Optional[str] = None
    created_by_member_id: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    # "<node_id>.<input_key>" for saved overrides the current workflow JSON no
    # longer accepts — these are dropped silently at dispatch. Computed per
    # response against the workflow on disk, never persisted.
    stale_overrides: List[str] = Field(default_factory=list)


class ReviewListResponse(BaseModel):
    items: List[ReviewRequestItem]
    total: int
    page: int
    pages: int
    # Rows per status across the whole queue (project/batch scope applies, the
    # `status` filter and pagination do not) — the Flow rail's stage counters
    # read this instead of counting the returned page. Statuses with no rows
    # are absent.
    status_counts: Dict[str, int] = Field(default_factory=dict)


class ReviewPatchRequest(BaseModel):
    prompt: Optional[str] = None
    settings: Optional[Dict] = None


class ReviewDispatchRequest(BaseModel):
    ids: List[str] = Field(min_length=1)


class ReviewRedispatchBulkRequest(BaseModel):
    ids: List[str] = Field(min_length=1)


class ReviewDispatchResponse(BaseModel):
    dispatched: List[str]
    skipped: List[str]

File: frontend/src/types/review.ts
export type ReviewProvider = 'kling' | 'comfy_video' | 'comfy_image'

export type ReviewStatus =
  | 'pending_review'
  | 'approved'
  | 'dispatched'
  | 'completed'
  | 'failed'
  | 'discarded'

export interface ReviewRequestItem {
  id: string
  batch_id: string
  source_image_path: string
  original_prompt: string
  prompt: string
  provider: ReviewProvider
  workflow_name: string | null
  settings: Record<string, unknown>
  status: ReviewStatus
  execution_id: string | null
  result_path: string | null
  error: string | null
  project_id?: string | null
  created_by_member_id?: string | null
  created_at: string | null
  updated_at: string | null
  // "<node_id>.<input_key>" for saved overrides the current workflow JSON no
  // longer accepts; they are dropped silently at dispatch.
  stale_overrides?: string[]
}

export interface ReviewListResponse {
  items: ReviewRequestItem[]
  total: number
  page: number
  pages: number
  // Rows per status across the whole queue — project scope applies, the status
  // filter and pagination do not. Statuses with no rows are absent, so read it
  // through `??  0`. The Flow rail counts from this, never from `items`, which
  // is capped at per_page.
  status_counts?: Partial<Record<ReviewStatus, number>>
}

export interface ReviewItemCreate {
  source_image_path: string
  prompt: string
  provider: ReviewProvider
  workflow_name?: string | null
  settings?: Record<string, unknown>
}

export interface ReviewCreateResponse {
  batch_id: string
  request_ids: string[]
}

export interface ReviewDispatchResponse {
  dispatched: string[]
  skipped: string[]
}
